feat: Adds port/ protocol support for host

walker233 opened the pull request:
This update extends the host resource to use correct
port and protocol for the redhat and debian families.
It uses a feature from bash and the timeout command
from the coreutils packages.

When port is specified but protocol is omitted, it
assumes tcp is what was intended. When neither port
nor protcol is specified it falls back to ping the
host.

Tries to address issue #1439 to some degree.

walker233 committed:
feat: Adds port/ protocol support for host

This update extends the host resource to use correct
port and protocol for the redhat and debian families.
It uses a feature from bash and the timeout command
from the coreutils packages.

When port is specified but protocol is omitted, it
assumes tcp is what was intended. When neither port
nor protcol is specified it falls back to ping the
host.

walker233 committed:
chore: Fixes linting issue

Appears that linting tests were failing.

chris-rock commented on lib/resources/host.rb:
```diff
@@ -121,12 +121,20 @@ def resolve(hostname)
+      # for those os's with timeout from coreutils packages
+      # TODO: Verify whether this works on other linux versions
+      if port && (inspec.os.redhat? || inspec.os.debian?)
+        resp = inspec.command("timeout 1 bash -c 'cat </dev/null >/dev/#{proto}/#{hostname}/#{port}'")
```
Should we check if bash is available? eg. via `command('bash').exist?`

walker233 replied:
That would probably be safe, but I thought for some reason it was nearly universal in inclusion. At least on redhat and debian families. Is that not the case. By the same token we are going to be safe with bash, we should probably do the same thing for the timeout command, which is in both redhat coreutils and debian coreutils packages. And were you suggesting we'd modify as:
~~~
if port && (inspect.os.redhat? || inspec.os.debian?) && 
   command('bash').exist? && command('timeout').exist?
~~~
Or were you thinking of a different modification.

chris-rock commented on lib/resources/host.rb:
```diff
@@ -121,12 +121,20 @@ def resolve(hostname)
-      ping.exit_status.to_i != 0 ? false : true
+    def ping(hostname, port = nil, proto = nil)
+      # assume tcp if port is given without corresponding protocol
+      if port && !proto
```
should we just set the default to `tcp`?
```
def ping(hostname, port = nil, proto = 'tcp')
```

walker233 replied:
I agree. That is a better way to do that.

chris-rock reviewed (commented):
Thank you @walker233 for this improvement!

walker233 committed:
refactor: Defaults protocol to tcp in ping method

This commit moves the setup for defaulting tcp as the protocol back into the method ping instead of having the extra internal checks.

Signed-off-by: Walker Wheeler <walker233@gmail.com>

walker233 committed:
test: Adds additional unit tests for host resource

Signed-off-by: Walker Wheeler <walker233@gmail.com>

walker233 committed:
test: Adds test cases for additional host parameters

Signed-off-by: Walker Wheeler <walker233@gmail.com>

walker233 committed:
test: Merges in new test cases

Signed-off-by: Walker Wheeler <walker233@gmail.com>

walker233 committed:
test: Removes invalid test case format.

Signed-off-by: Walker Wheeler <walker233@gmail.com>

walker233 committed:
test: Fixes tests to call host resource correctly

Signed-off-by: Walker Wheeler <walker233@gmail.com>

walker233 committed:
test: Fixes miss typed port number

Signed-off-by: Walker Wheeler <walker233@gmail.com>

walker233 committed:
test: Changes port for unreachable test

Signed-off-by: Walker Wheeler <walker233@gmail.com>
